From bef54a59715bce18be40201d40429b973d377af0 Mon Sep 17 00:00:00 2001
From: Daniel Golle <daniel@makrotopia.org>
Date: Thu, 16 Mar 2023 16:32:42 +0530
Subject: [PATCH 045/500] OpenWrt:
 493-ubi-set-ROOT_DEV-to-ubiblock-rootfs-if-unset.patch

ubi: set ROOT_DEV to ubiblock "rootfs" if unset

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

Change-Id: Id281c5f99db85a661707e6be74e7a1e2a14ab61c
Signed-off-by: Ram Chandra Jangir <quic_rjangir@quicinc.com>
(cherry picked from commit 985bdfe095b0d941087961874ba14fc0154a241b)
---
 drivers/mtd/ubi/block.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/mtd/ubi/block.c b/drivers/mtd/ubi/block.c
index 2d734d83faa1..d0360d572eef 100644
--- a/drivers/mtd/ubi/block.c
+++ b/drivers/mtd/ubi/block.c
@@ -41,6 +41,7 @@
 #include <linux/scatterlist.h>
 #include <linux/idr.h>
 #include <asm/div64.h>
+#include <linux/root_dev.h>
 
 #include "ubi-media.h"
 #include "ubi.h"
@@ -428,6 +429,15 @@ int ubiblock_create(struct ubi_volume_info *vi)
 	dev_info(disk_to_dev(dev->gd), "created from ubi%d:%d(%s)",
 		 dev->ubi_num, dev->vol_id, vi->name);
 	mutex_unlock(&devices_mutex);
+
+	if (!strcmp(vi->name, "rootfs") &&
+		IS_ENABLED(CONFIG_MTD_ROOTFS_ROOT_DEV) &&
+		ROOT_DEV == 0) {
+		    pr_notice("ubiblock: device ubiblock%d_%d (%s) set to be root filesystem\n",
+			      dev->ubi_num, dev->vol_id, vi->name);
+		ROOT_DEV = MKDEV(gd->major, gd->first_minor);
+	}
+
 	return 0;
 
 out_remove_minor:
-- 
2.34.1

